Fixed dummy task based on devp2p branch

maaktweluit opened the pull request:
Resolves #1443

mfranciszkiewicz commented on tests/golem/task/dummy/runner.py:
```diff
@@ -263,13 +262,13 @@ def dispatch(args):
-        fail_after = float(args[4]) if len(args) == 5 else None
-        run_computing_node(args[2], SocketAddress.parse(args[3]), fail_after=fail_after)
+        fail_after = float(args[5]) if len(args) == 6 else None
+        run_computing_node(args[2], SocketAddress.parse(args[3]), args[4], fail_after=fail_after)
```
`node_num` is never passed via command line (line 195). As a matter of fact, it is not needed in this version.

maaktweluit replied:
`node_num` is required to increment the port, it ensures unique hyperg:port instances can run at the same time.

I see it is not passed on line 195, but the arguments seem lower then 5 there, so it will not trigger this `elif`.

What would be your proposed solution?
- Add `node_num` to 195?
- Remove the need for `node_num`, solving the ports differently
- .... ( anything i could not think of? )

mfranciszkiewicz replied:
`develop`, unlike `devp2p`, tries to bind to the next free port if current one is taken. There is no need to manually manage ports. And currently, `hyperg`'s port is not set by Golem.

This file could probably be left unchanged. The removal of `unittest.skip` and changes in DummyTask itself should suffice.

It's up to you whether you want to:
- add `node_num` and have better control over assigned ports
- revert this file and keep the changes to a minimum

maaktweluit replied:
Nice, thanks for the clear answer!

Will try keeping changes to a minimum if possible, remember errors about hyperg ports but they could be from elsewhere.

maaktweluit replied:
reverted the 2 files back, the first computing tests now fails without clear error in the logs:

Remember digging in and finding an error on ports somewhere, will try to find it again.
Where should it pick the ports automatically? maybe i can add some logging there to confirm.
```
E       AssertionError: 'Computation timed out' not found in 'Node exit
ed with return code 0'

tests/golem/task/dummy/test_dummy_task.py:14: AssertionError
------------------------ Captured stdout call -------------------------
[REQUESTOR 10011] Starting in /var/folders/sy/873ntd415m53f3n9h86dchlh0
000gn/T/golem_dummy_simulation_jvp9dgf0/requestor
2017-09-17 22:28:57.600 [HyperG] [INFO] HyperG [0.2.1]
2017-09-17 22:28:57.602 [HyperG] [INFO] TCP listening on 0.0.0.0:3282
[REQUESTOR 10011] Started in 4.0 s
[REQUESTOR 10011] Listening on 192.168.1.10:40102
[COMPUTER  10047] Starting in /var/folders/sy/873ntd415m53f3n9h86dchlh0
000gn/T/golem_dummy_simulation_jvp9dgf0/computer1
[COMPUTER  10046] Starting in /var/folders/sy/873ntd415m53f3n9h86dchlh0
000gn/T/golem_dummy_simulation_jvp9dgf0/computer0
2017-09-17 22:28:59.529 [HyperG] [INFO] Sharing 0cc599822f0c0301c97429f
a0b880b720bf3d56c8934f181132a2d52974a1e28
[COMPUTER  10046] Started in 2.2 s
[COMPUTER  10046] Connecting to requesting node at 192.168.1.10:40102 .
..
[COMPUTER  10047] Started in 2.2 s
[COMPUTER  10047] Connecting to requesting node at 192.168.1.10:40102 .
..
Stopping nodes...
------------------------ Captured stderr call -------------------------
create HyperG logsdir: /var/folders/sy/873ntd415m53f3n9h86dchlh0000gn/T
/golem_dummy_simulation_jvp9dgf0/requestor/logs
Unhandled error in Deferred:

create HyperG logsdir: /var/folders/sy/873ntd415m53f3n9h86dchlh0000gn/T
/golem_dummy_simulation_jvp9dgf0/computer1/logs
create HyperG logsdir: /var/folders/sy/873ntd415m53f3n9h86dchlh0000gn/T
/golem_dummy_simulation_jvp9dgf0/computer0/logs
```

mfranciszkiewicz replied:
`BASE_PORT` is used for Golem's ports, not hyperg. 

Every `Client` checks if a hyperg instance is already running. The problem here is that requestor and computers try to start their own instance of hyperg (almost) at the same time. The first one succeeds, the others fail because port 3282 is taken. Typical race condition.

There is a possibility to change listening and rpc ports in hyperg (try `--help`), but it is not currently possible to do so with Golem. 

Ways to resolve this:
- computers spawn only when requestor's hyperg is running
- hyperg is running in the background, prior to starting the test
- an option to disable hyperg daemon manager; used in computers
- an option to define custom hyperg ports in `Client` and daemon manager; a command line arg for `golemapp.py` would also be nice
- spawning computers is delayed (e.g. sleep?)

maaktweluit replied:
tried a few of the options:
- Hyperg already running ( this is also the case on CI ) did not make any difference
- Sleep before spawning computers, beginning of `for`, did not make a difference

Will go back to the working version for me, propose to make follow up issues for the proposed improvements.

Should I do anything with the `node_num` on line 195? Looking at the devp2p branch there `n` is added to the command.

maaktweluit replied:
when adding `str(n)` to 195 it stopped working..

Will look into your other proposed options tomorrow :)

did find the error again that was causing this:
```
{"message":"listen EADDRINUSE 0.0.0.0:3282","stack":"Error: lis
ten EADDRINUSE 0.0.0.0:3282\n    at Object.exports._errnoException (uti
l.js:1022:11)\n    at exports._exceptionWithHostPort (util.js:1045:20)\
n    at Server._listen2 (net.js:1259:14)\n    at listen (net.js:1295:10
)\n    at net.js:1405:9\n    at _combinedTickCallback (internal/process
/next_tick.js:77:11)\n    at process._tickCallback (internal/process/ne
xt_tick.js:98:9)\n    at Module.runMain (evalmachine.<anonymous>:0)\n
  at run (bootstrap_node.js:398:7)\n    at startup (bootstrap_node.js:1
53:9)","code":"EADDRINUSE","errno":"EADDRINUSE","syscall":"listen","add
ress":"0.0.0.0","port":3282}
```
it is strange how it does not consistently show up in the error logs.

mfranciszkiewicz replied:
So the problem lies somewhere else, still hyperg will be an issue.

Port binding:
https://github.com/golemfactory/golem/blob/b0506b28a4b548f1ff315332c82cfe935f003470/golem/network/transport/tcpnetwork.py#L246-L254 `...`
https://github.com/golemfactory/golem/blob/b0506b28a4b548f1ff315332c82cfe935f003470/golem/network/transport/tcpnetwork.py#L371-L378

maaktweluit replied:
thank you, will dig in and get to the real problem!

maaktweluit commented on tests/golem/task/dummy/runner.py:
```diff
@@ -263,13 +262,13 @@ def dispatch(args):
         # I'm a computing node,
         # second arg is the data dir,
         # third arg is the address to connect to,
         # forth arg is the timeout (optional).
```
comment out of date, will fix it when `node_num` is sorted

maaktweluit commented:
After a lot of debugging and comparing `0.8.0` with this version the errors nor ports matter.

The first line missing from the current branch in the logs is hyperg "looking up" the resources. @Radagast-red do you know where this ( accept task and download resource ) should be happening?

Calling it a day for now, will continue with a fresh look :)

Radagast-red commented:
Downloading resource:
https://github.com/golemfactory/golem/blob/develop/golem/network/hyperdrive/client.py#L61
I am not sure what do you mean by accepting task, probably you can start here:
https://github.com/golemfactory/golem/blob/develop/golem/task/taskserver.py#L184
Here golem takes one of known task descriptors (aka TaskHeader) and tries to connect directly to task owner. When it does and receives: MessageResourceList
It will pull resources:
https://github.com/golemfactory/golem/blob/develop/golem/task/tasksession.py#L670

maaktweluit commented:
Found and fixed it!
Turned out the price was not allowed to be equal, only greater then.
Also added `config_logging` to the dummy nodes so on errors there is logging available.

@Radagast-red @mfranciszkiewicz Thank you for all the help and please review again.

Propose the following "follow up" tasks:
- [ ] Add better verification to the dummy task ( None | exit code 0 is minimal )
- [ ] Remove old dictionary ( `isinstance` ) based code facilitating the dummy task.
- [ ] Add `task_defenition` to dummy task so there are no ( ignorable ) errors

@badb  Please advice on what issue to make where, if any.

maaktweluit commented on tests/golem/task/dummy/test_runner_script.py:
```diff
@@ -1,12 +1,11 @@
 
 from golem.network.transport.tcpnetwork import SocketAddress
 from golem.testutils import DatabaseFixture
+from golem.tools.ci import ci_skip
```
unused var :(

Elfoniok commented on tests/golem/task/dummy/task.py:
```diff
@@ -187,9 +189,10 @@ def verify_task(self):
     def verify_subtask(self, subtask_id):
-        result = self.subtask_results[subtask_id]
+        from golem.core.simpleserializer import CBORSerializer
+        result = CBORSerializer.loads(self.subtask_results[subtask_id])
```
Why serialization here?

maaktweluit replied:
was taken from `devp2p`, will remove if it was only needed there

maaktweluit replied:
done

Elfoniok commented on tests/golem/task/dummy/task.py:
```diff
@@ -66,7 +67,7 @@ def __init__(self, client_id, params, num_subtasks):
             resource_size=params.shared_data_size + params.subtask_data_size,
             estimated_memory=0,
-            max_price=MIN_PRICE)
+            max_price=MIN_PRICE, docker_images=[])
```
Are you sure you want to init this with [ ] (it will be same list every function call), instead None?

maaktweluit replied:
This was taken from the `devp2p` branch, will check if the called function will handle it well and update it

maaktweluit replied:
done

jiivan commented on tests/golem/task/dummy/test_runner_script.py:
```diff
@@ -1,12 +1,10 @@
-import unittest
-
 import mock
```
Since you're editing this test: pls use stdlibs `unittest.mock`.

maaktweluit replied:
Will do! though the interface will change a bit i've seen.. i'll give it a shot!

maaktweluit replied:
done
